drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+	set_plane_property(req, plane, dev->plane_id, "SRC_W", buf->width << 16);
+	set_plane_property(req, plane, dev->plane_id, "SRC_H", buf->height << 16);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_X", 0);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_Y", 0);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_W", buf->width);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_H", buf->height);
+
+	/* Set the connector, CRTC and plane properties in a single IOCTL. Here we
+	 * leave it as a blocking call, since we can't start to do page-flips before
+	 * setting everything up.
+	 */
+	flags = DRM_MODE_ATOMIC_ALLOW_MODESET;
+	ret = drmModeAtomicCommit(fd, req, flags, NULL);
+
+	drmModeAtomicFree(req);
+
+	return ret;
+}
+
+/*
+ * Just like modeset_atomic_modeset(), in modeset_atomic_commit() we're going to
+ * atomically set some properties for our objects. The difference is that here
+ * we are setting up just the planes properties. So this is used when we want to
+ * flip a page and not during the modeset setup.
+ */
+
+static int modeset_atomic_commit(int fd, struct modeset_buf *buf, struct modeset_dev *dev)
+{
+	struct drm_object *plane = &dev->plane;
+	drmModeAtomicReq *req;
+	int ret, flags;
+
+	req = drmModeAtomicAlloc();
+
+	set_plane_property(req, plane, dev->plane_id, "FB_ID", buf->fb);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_ID", dev->crtc_id);
+	set_plane_property(req, plane, dev->plane_id, "SRC_X", 0);
+	set_plane_property(req, plane, dev->plane_id, "SRC_Y", 0);
+	set_plane_property(req, plane, dev->plane_id, "SRC_W", buf->width << 16);
+	set_plane_property(req, plane, dev->plane_id, "SRC_H", buf->height << 16);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_X", 0);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_Y", 0);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_W", buf->width);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_H", buf->height);
+
+	/* we want to receive the page-flip event and we want this
+	 * call to be non-blocking */
+	flags = DRM_MODE_PAGE_FLIP_EVENT | DRM_MODE_ATOMIC_NONBLOCK;
+	ret = drmModeAtomicCommit(fd, req, flags, dev);
+
+	drmModeAtomicFree(req);
+
+	return ret;
+}
+
+/*
+ * modeset_draw() changes just a little bit. Now we have to call the function
+ * to setup the modeset for each output.
+ * 
+ * We also reinforce that here the draw happens for the first time. In the
+ * not-atomic vsync'ed version, the KMS setup was done using drmModeSetCrtc(),
+ * which would also cause a page-flip. Using the atomic API this doesn't happen.
+ */
+
+static void modeset_draw(int fd)
+{
+	int ret;
+	fd_set fds;
+	time_t start, cur;
+	struct timeval v;
+	drmEventContext ev;
+	struct modeset_dev *iter;
+
+	/* initial modeset on all outputs; note that this is done with a blocking
+	 * call, since we can't do page-flips before setting KMS up */
+	for (iter = modeset_list; iter; iter = iter->next) {
+		modeset_atomic_modeset(fd, iter);
```

Comment:
Create one `drmModeAtomicReq` before this loop. Pass it into the function inside the loop. Do the atomic commits after the loop. This gets all the outputs into the same initial commit like they should.

Reply:
I mean, do a test-only atomic commit here, and then the real atomic commit after you have painted the framebuffers.

Reply:
Done.